feat: introduction of simple table functions

benbellick opened the pull request:
- feat: introduces a notion of YAML-defined table functions which are represented as a new `TableFunctionRel` (take in no relations)

Closes #823

benbellick committed:
feat: introduction of simple table functions

introduces a notion of YAML-defined functions which are represented as
a new `TableFunctionRel`.

Closes #823

benbellick commented on proto/substrait/algebra.proto:
```diff
@@ -552,6 +552,51 @@ message ExpandRel {
+  //    string that case-insensitively matches one of the allowed options.
+  repeated FunctionArgument arguments = 3;
+
+  // The derived fields indicates whether or not the YAML file produced the schema:
```
Every other function type has a `FunctionOption` field around here. Do we need that for table functions as well? 
```protobuf
message TableFunctionRel {
    ...
    repeated FunctionOption options = 4;
    ...
}
```

yongchul replied:
For the sake of consistency, maybe it is worth it? I don't understand what this FunctionOption is for... (code archaeology time!)

benbellick commented on text/simple_extensions_schema.yaml:
```diff
@@ -307,3 +311,33 @@ $defs:
+              $ref: "#/$defs/sessionDependent"
+            deterministic:
+              $ref: "#/$defs/deterministic"
+            schema:
```
Would it be better to come up with a more explicit way to mark the inability of deriving an output schema for a particular table function? Or is it sufficient to just let the absence of the `schema` field indicate this?

benbellick commented on proto/substrait/function.proto:
```diff
@@ -106,6 +106,19 @@ message FunctionSignature {
     }
   }
 
+  message TableFunction {
```
I added this to be consistent with the other function definitions, though AFAICT, this isn't really used anywhere. I found #193 from a few years ago which suggests that this should continue to be updated, but I'm not sure if that is still the prevailing philosophy.

benbellick commented on proto/substrait/function.proto:
```diff
@@ -106,6 +106,19 @@ message FunctionSignature {
+
+    NamedStruct schema = 9;
+
+    repeated Implementation implementations = 10;
```
Other functions have variadic final argument logic like: 
```protobuf
message TableFunction {
    ...
    oneof final_variable_behavior {
      FinalArgVariadic variadic = 10;
      FinalArgNormal normal = 11;
    }
    ...
}
```

However, I skipped it for now for simplicity. Is it necessary on a first pass?

benbellick committed:
fix: table functions output as return instead of schema

This is for consistency with other functions

benbellick committed:
wip: fix doc generation for absent return type (table functions)

benbellick commented on site/docs/extensions/generate_function_docs.py:
```diff
@@ -125,19 +125,27 @@ def write_markdown(file_obj: dict, file_name: str) -> None:
                 # If the return value for the function implementation is multiple lines long,
                 # print each line separately. This is the case for some functions in
                 # functions_arithmetic_decimal.yaml
-                if "\n" in impl["return"]:
```
TODO: this could probably use a little work to make it a bit nicer. I'll work on that later once we have some general agreement on the spec changes. 

The change here is for now just to get CI/CD to pass.

benbellick committed:
docs: small improvements to docs for table functions

yongchul commented on extensions/functions_table.yaml:
```diff
@@ -0,0 +1,102 @@
+            types:
+              - i32
+  - name: "unnest"
+    description: Expands a list literal into a set of rows, one row per element.
```
list expression? we don't have to restrict it to a literal

yongchul commented on proto/substrait/algebra.proto:
```diff
@@ -552,6 +552,51 @@ message ExpandRel {
+//
+// Future extensions may add an optional input field to support transformation
+// table functions that operate on input relations.
+message TableFunctionRel {
```
From my biased view, it is better to categorize them... roughly

* Scan like (0 relational input)
* Project or Filter like (1 relational input)
* (N-ary) Join like (2+ relational inputs optional grouping support)
* Group like (1 relational input with grouping support)

This gives way more modular and systematic way to describe a table function.

yongchul replied:
I prefer `Rel` for each type, clearly expressing the degree of input at message level. But I'm open to exploring `oneof` definition.

yongchul reviewed (commented):
Flushing some comments as my meetings getting longer and have on call duty coming up in the afternoon. Will come back later today.

yongchul commented on extensions/functions_table.yaml:
```diff
@@ -0,0 +1,102 @@
+    impls:
+      - args:
+          - name: input
+            value: "list<T>"
```
perhaps we can follow other extensions. for instance, in decimal rounding functions, we just write type rather than wrap it in a quote (i.e., decimal<P,S> rather than "decimal<P,S>").

Also, I'm wondering whether we can follow the same pattern as others (i.e., allow type expression).

benbellick committed:
fix: fix some of @yongchul's comments

benbellick committed:
tweak: drop session dependence and deterministic fields from rel

yongchul commented on extensions/functions_table.yaml:
```diff
@@ -0,0 +1,92 @@
+            description: The starting value of the series
+          - name: stop
+            value: i64
+            description: The ending value of the series (inclusive)
```
Looks like there is an interesting spilit between `stop` being `inclusive` and `exclusive`.

* python, pandas and snowflake exclude the stop value.
* PostgreSQL, T-SQL, ZetaSQL include the stop value.

I guess either way is fine... Another problem is that some systems do support generating series for other numeric types (e.g., fp32, fp64, decimal) but we can add those later.

yongchul commented on extensions/functions_table.yaml:
```diff
@@ -0,0 +1,92 @@
+
+      Takes constant arguments and produces zero or more records containing a single
+      integer value. The series includes both the start and stop values if they fall
+      on a step boundary. If step is positive, stops when the value exceeds stop.
```
Do we want to say something about the default value of `step`?

benbellick commented on site/docs/expressions/table_functions.md:
```diff
@@ -1,8 +1,265 @@
+
+- **Transformation table functions**: Functions that take an input relation and transform it (by adding an optional `input` field to `TableFunctionRel`)
+- **Set-returning functions**: Functions that process input records and produce multiple output records per input
+- **Lateral joins**: Applying table functions to each row of an input relation
```
At least in postgresql, `SELECT unnest(col) FROM table;` is legal (see example [here](https://www.db-fiddle.com/f/cEaSifnLbWpZESAvx3xtJP/1)). This is documented [here](https://www.postgresql.org/docs/current/xfunc-sql.html#XFUNC-SQL-FUNCTIONS-RETURNING-SET). 
> Functions returning sets can also be called in the select list of a query. For each row that the query generates by itself, the set-returning function is invoked, and an output row is generated for each element of the function's result set. 

and also
> PostgreSQL's behavior for a set-returning function in a query's select list is almost exactly the same as if the set-returning function had been written in a LATERAL FROM-clause item instead. For example,
```sql
SELECT x, generate_series(1,5) AS g FROM tab;
```
> is almost equivalent to

```sql 
SELECT x, g FROM tab, LATERAL generate_series(1,5) AS g;
```
> It would be exactly the same, except that in this specific example, the planner could choose to put g on the outside of the nested-loop join, since g has no actual lateral dependency on tab. That would result in a different output row order. Set-returning functions in the select list are always evaluated as though they are on the inside of a nested-loop join with the rest of the FROM clause, so that the function(s) are run to completion before the next row from the FROM clause is considered.

Though of course, postgresql is just one DB in a sea of DBs 😅

yongchul replied:
Hmm... I see. Looks like PostgreSQL chose to allow the syntactic sugar (to me) which I disagree to have but oh well. 😆

benbellick commented on extensions/functions_table.yaml:
```diff
@@ -0,0 +1,92 @@
+          - name: step
+            value: i64
+            description: The increment between values
+            constant: true
```
https://substrait.io/expressions/scalar_functions/#argument-types
> Value arguments: arguments that refer to a data value. These could be constants (literal expressions defined in the plan) or variables (a reference expression that references data being processed by the plan).

My understanding is that `constant` means that a value can only be a literal and not a reference to a value unknown at plan construction e.g. a column reference. Although I'm unsure why we wouldn't allow these values to be expressions which lack references. 

All that being said, postgresql does in fact accept column references in the arguments to generate_series (example [here](https://www.db-fiddle.com/f/cgG7YG1UWeFxEL2pACzNkT/0)), so I think it makes sense to lift this restriction.

yongchul replied:
We shouldn't unnecessarily restrict ourselves to literals and this is another place we shouldn't IMO. Implementation may not support generic expression in certain places but we shouldn't prevent it from the spec. In fact, this is another potential information to encode in **dialect** whether an expression has some restrictions.

benbellick committed:
fix: address @yongchul's comments

- drop restriction to constant args in table functions
- drop mentions of "runtime" dependency
- drop derived field from proto message for table functions

benbellick commented on proto/substrait/algebra.proto:
```diff
@@ -552,6 +552,61 @@ message ExpandRel {
+  // this must match the schema computed by evaluating the YAML ExpressionNamedStruct
+  // with the bound argument types. For runtime-dependent schemas (derived=false),
+  // this is provided by the plan producer.
+  NamedStruct table_schema = 7;
```
Leaving this conversation unresolved if others have something to say, but my most recent commit has dropped `derived` from the proto. Thanks!

benbellick committed:
docs: consolidate redundant examples

benbellick committed:
feat: add variadic handling to table functions

benbellick committed:
cleanup: make some adjustments

- use any1 instead of T
- add variadic-ness to unnest and to table functions in general

benbellick commented on extensions/functions_table.yaml:
```diff
@@ -0,0 +1,93 @@
+            description: The list(s) to unnest
+        variadic:
+          min: 1
+        return:
```
I have made this a draft PR again. I need to have a proper think about how we can represent this dynamic return type if we really want to have these functions be variadic 😰
